server/aesdsocket.c

```diff
@@ -0,0 +1,372 @@
+
+        case STATE_LISTENING:
+            // LOG_DBG("STATE_LISTENING\n");
+            //listen for connections
+            status = listen(sockfd_server, 5);
+            if (status < 0)
+            {
+                perror("Listening failed");
+                return_status = -1;
+                server_socket_state = EXIT;
+                break;
+            }
+            server_socket_state = STATE_OPENING_FILE;
+            break;
+
+        case STATE_OPENING_FILE:
+        //    LOG_DBG("STATE_OPENING_FILE\n");
+           fd = open("/var/tmp/aesdsocketdata", O_CREAT | O_RDWR | O_TRUNC, 0644);
+            if (fd == -1)
+            {
+                perror("error creating file");
+                return_status = -1;
+                server_socket_state = EXIT;
+                break;
+            }
+            server_socket_state = STATE_ACCEPTING;
+            break;
+
+        case STATE_ACCEPTING:
+            // LOG_DBG("STATE_ACCEPTING\n");
+            addr_size = sizeof(new_addr);
+		
+            //accept socket connections
+            sockfd_client = accept(sockfd_server, (struct sockaddr*)&new_addr, &addr_size);
+            if(sockfd_client < 0) 
+            {
+                LOG_ERROR("error accepting connection");
+                server_socket_state = EXIT;
+                break;
+            }
+
+            inet_ntop(AF_INET, &client_addr, str, INET_ADDRSTRLEN);
+            
+		    syslog(LOG_INFO, "Accepted connection from %s", str);
+
+
+            nwrite=0, nreceived=0, nsent=0, nread=0;
+            heap_buffer_for_write = NULL;
+            heap_buffer_for_read = NULL;
+            filled_up_buffer_size = 0;
+            total_buffer_size = BUF_SIZE;
+
+            heap_buffer_for_write = (char *)malloc(sizeof(char)*BUF_SIZE);
+            if(heap_buffer_for_write == NULL) 
+            {
+                LOG_ERROR("could not malloc");
+                server_socket_state = EXIT;
+                break;
+            }
+            server_socket_state = STATE_RECEIVE_FROM_CLIENT;
+            break;
+
+        case STATE_RECEIVE_FROM_CLIENT:
+            // LOG_DBG("STATE_RECEIVE_FROM_CLIENT\n");
+            do {
+			nreceived = recv(sockfd_client, buf, sizeof(buf), 0);
+			//LOG_DBG("Received %d\n", nreceived);
+			if(!nreceived || (strchr(buf, '\n')!=NULL))
+				server_socket_state = STATE_WRITING_FILE;
+
+			if((total_buffer_size - filled_up_buffer_size) < nreceived) {
+
+				total_buffer_size += nreceived;
+				heap_buffer_for_write = (char *)realloc(heap_buffer_for_write, sizeof(char)*total_buffer_size);
+			}
+			
+			memcpy(&heap_buffer_for_write[filled_up_buffer_size], buf, nreceived);
+			filled_up_buffer_size += nreceived;
+			
+		    } while(server_socket_state == STATE_RECEIVE_FROM_CLIENT);
+
+            break;
+
+        case STATE_WRITING_FILE:
+            // LOG_DBG("STATE_WRITING_FILE\n");
+            nwrite = write(fd, heap_buffer_for_write, filled_up_buffer_size);
```

Comment:
@rajatchaple what assumption does this make about the total size of all packets sent relative to the size of the heap?  What happens if 1GB of packets were sent?